### PR TITLE
Fix command route inheritance

### PR DESCRIPTION
**Pull request by timbussmann**
Connects to Particular/V6Launch#68

We no longer look for routes from base classes when resolving routes for commands as concluded in https://github.com/Particular/NServiceBus/issues/3911

I split the send and the publish router since they no longer share all of the logic and it may become easier to optimize the specific routers.

@Particular/nservicebus-maintainers @DavidBoike @bording @SzymonPobiega @Scooletz please review

### CODE REVIEW

**Comment by danielmarbach** [comment]
@timbussmann Is this PR self contained so that it can be merged although we want to release beta 6 tomorrow morning? Or should we postpone the merge of this one?

**Comment by timbussmann** [comment]
that should go together with https://github.com/Particular/NServiceBus/pull/3979 I think.

I'm ok with merging that after release beta6 so we can focus beta7 on the routing changes.

**Comment by danielmarbach** [comment]
Can we make those WIP then?

**Comment by timbussmann** [comment]
@danielmarbach I'd rather not. They are ready for review. If for some reason we don't want to merge them before tomorrow, let's just keep them open?

**Comment by SzymonPobiega** [comment]
I see there is a lot of duplication between the new router classes. Can you elaborate on the purpose of getting rid of the base class?

**Comment by danielmarbach** [comment]
@timbussmann sorry now this needs a rebase :(

**Comment by timbussmann** [comment]
> I see there is a lot of duplication between the new router classes. Can you elaborate on the purpose of getting rid of the base class?

there is quite a bit of duplication, but it's the details which are different. Also I didn't have a closer look whether we can further optimize the specific routers to reduce hot path execution. When all issues for the routing are resolved and we still see the routers being nearly identical, we can still easily merge them again, while optimizing a generic/abstract class is generally much more difficult.

**Comment by timbussmann** [comment]
rebased

**Comment by SzymonPobiega** [comment]
I did a comparison and it seems like only the `GetDestinations` method contains the difference. By extracting one method I was able to make `Route` and `SelectDestinationsForEachEndpoint` be exactly the same so I think deduplicating it is worth the effort.

**Comment by timbussmann** [comment]
@SzymonPobiega are you ok with adding this as a task after potential hot path optimizations (after all design changes have been applied) instead? So if the code remains this way we can deduplicate again but it won't block us from finding optimizations.

**Comment by SzymonPobiega** [comment]
I don't think I agree. I see more value in code being clean then being super fast. When is an opportunity to reduce duplication, it should be done first and then, if we have proofs that large performance optimization can occur, it might be split up again to exploit a specific performance hack.

**Comment by timbussmann** [comment]
> I don't think I agree. I see more value in code being clean then being super fast.

@SzymonPobiega the maintainers decided clearly that we value hot path performance over unnecessary code and allocations (e.g. linqs, ICollection interfaces and so on). Deduplicating command and publishing routing logic falls under the same category for me. Especially since we're talking about further changes to the routing table (e.g. https://github.com/Particular/NServiceBus/issues/3909 you and me are discussing how to make routing table essentially return a single result instead of a collection which will directly affect the router). We're moving towards the final routing code in small PRs and usually, the refactoring comes last when we're happy with the behavior itself. Publishing and sending a command are in my opinion more than different enough to justify dedicated classes if that turns out to be even slightly more performant. I'd rather see subscription and command routing be deducplicated, but even those two routers are too different in their behavior that it currently doesn't make sense. /cc @danielmarbach

**Comment by danielmarbach** [comment]
@SzymonPobiega what @timbussmann said

File: src/NServiceBus.Core.Tests/Routing/Routers/UnicastSendRouterConnectorTests.cs [hunk 1]
```diff
@@ -166,14 +166,14 @@ public class UnicastSendRouterConnectorTests
         [Test]
         public async Task Should_route_using_the_mappings_if_no_destination_is_set()
         {
-            var strategy = new FakeRoutingStrategy
+            var strategy = new FakeSendRouter
```
Comment (bording) [inline src/NServiceBus.Core.Tests/Routing/Routers/UnicastSendRouterConnectorTests.cs]: nitpick: `strategy` should probably be renamed to `router` or something. Same for the other tests in this file.

File: src/NServiceBus.Core/Routing/Routers/UnicastSendRouterConnector.cs [hunk 2]
```diff
@@ -22,7 +22,7 @@ public enum RouteOption
         public UnicastSendRouterConnector(
             string sharedQueue,
             string instanceSpecificQueue,
-            IUnicastRouter unicastRouter,
+            IUnicastSendRouter unicastRouter,
```
Comment (bording) [inline src/NServiceBus.Core/Routing/Routers/UnicastSendRouterConnector.cs]: I'd rename this and the field to `unicastSendRouter` to match the change to the `UnicastPublishRouterConnector `.

**Comment by bording** [comment]
Aside from my few renaming comments, this looks good to me! :+1:

File: src/NServiceBus.Core.Tests/Routing/Routers/UnicastSendRouterConnectorTests.cs (hunk 1 above)
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Routing/Routers/UnicastSendRouterConnectorTests.cs]: good catch, missed that

**Comment by timbussmann** [comment]
addressed @bording comments. Ready to go.

File: src/NServiceBus.Core.Tests/Routing/RoutingSettingsTests.cs [hunk 3]
```diff
@@ -148,6 +131,14 @@ static UnicastRoutingTable ApplyConfiguredRoutes(RoutingSettings routingSettings
             }))))).SelectMany(x => x);
         }
 
+        static Task<IEnumerable<UnicastRoutingTarget>> RetrieveRoutingTarget(IUnicastRoute result)
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Routing/RoutingSettingsTests.cs]: is this even used? Quick search on GH UI returned nothing

**Comment by danielmarbach** [comment]
@timbussmann triggered the build again. Made one comment

File: src/NServiceBus.Core.Tests/Routing/RoutingSettingsTests.cs (hunk 3 above)
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Routing/RoutingSettingsTests.cs]: good catch! that code became obsolete. Removed it

**Comment by timbussmann** [comment]
@danielmarbach addressed your finding :+1:

**Pull request merged**